### PR TITLE
Add test cases for blitFramebuffe to cover more corner cases.

### PR DESCRIPTION
**Pull request by Richard-Yunchao**
If the src buffer and dst buffer are identical, generates INVALID_OPERATION.
See the gles 3.0.4 spec about this at https://www.khronos.org/registry/gles/specs/3.0/es_spec_3.0.4.pdf#nameddest=section-4.3.3. It is at the middle on page 198.
 
On the gles3 man page of blitFramebuffer at https://www.khronos.org/opengles/sdk/docs/man3/html/glBlitFramebuffer.xhtml. It says that this behavior is not defined. However, gles3 man page is not always updated. Moreover, to be undefined is not a good definition.

PTAL. Thanks a lot!

### CODE REVIEW

File: specs/latest/2.0/index.html [hunk 1]
```diff
@@ -1273,7 +1273,11 @@ <h2 class="no-toc">Table of contents</h2>
         Transfer a rectangle of pixel values from one region of the read framebuffer to another in the draw
         framebuffer. If the value of SAMPLE_BUFFERS for the read framebuffer is one and the value of
         SAMPLE_BUFFERS for the draw framebuffer is zero, the samples corresponding to each pixel location in
-        the source are converted to a single sample before being written to the destination.
+        the source are converted to a single sample before being written to the destination. <br><br>
+
+        If the source and destination buffer are identical, which means they have the same image with the same mipmap level
+        of texture and the same layer of 3D or 2D_ARRAY texture and the same face of CUBE_MAP texture,
+        generates an <code>INVALID_OPERATION</code> error.
```
Comment (zhenyao) [inline specs/latest/2.0/index.html]: I don't think this is necessary, because there is no deviation from the ES3 spec.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        m_data0[ii * 4 + 3] = 0xff;
+    }
+    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA8, width, height, 0, gl.RGBA, gl.UNSIGNED_BYTE, m_data0);
+    gl.generateMipmap(gl.TEXTURE_2D);
+
+    var fb0 = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.READ_FRAMEBUFFER, fb0);
+    gl.framebufferTexture2D(gl.READ_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex_2d, 0);
+    if (gl.checkFramebufferStatus(gl.READ_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE) {
+        testFailed("Read framebuffer incomplete. Fail reason: " + wtu.glEnumToString(gl, status));
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: status is undefined.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    // Create draw fbo and its color attachment.
+    var rb0 = gl.createRenderbuffer();
+    gl.bindRenderbuffer(gl.RENDERBUFFER, rb0);
+    gl.renderbufferStorage(gl.RENDERBUFFER, gl.RGBA8, width, height);
+
+    var fb1 = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fb1);
+    gl.framebufferRenderbuffer(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.RENDERBUFFER, rb0);
+    if (gl.checkFramebufferStatus(gl.DRAW_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE) {
+        testFailed("Draw framebuffer incomplete. Fail reason: " + wtu.glEnumToString(gl, status));
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: status is undefined.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    gl.bindFramebuffer(gl.READ_FRAMEBUFFER, fb0);
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fb0);
+    gl.blitFramebuffer(0, 0, 2, 2, 4, 4, 6, 6, gl.COLOR_BUFFER_BIT | gl.COLOR_BUFFER_BIT, gl.NEAREST);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "blitFramebuffer should generate INVALID_OPERATION if read/draw buffer are identical.");
+
+    // The same image with the same level bound to read/draw buffer.
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fb1);
+    gl.framebufferTexture2D(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex_2d, 0);
+    if (gl.checkFramebufferStatus(gl.DRAW_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE) {
+        testFailed("Draw framebuffer incomplete. Fail reason: " + wtu.glEnumToString(gl, status));
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: status is undefined.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        return;
+    }
+    gl.blitFramebuffer(0, 0, 2, 2, 4, 4, 6, 6, gl.COLOR_BUFFER_BIT | gl.COLOR_BUFFER_BIT, gl.NEAREST);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "blitFramebuffer should generate INVALID_OPERATION if read/draw buffer are identical.");
+
+    // The same image in read/draw buffer, but different levels are bound to read/draw buffer respectively.
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fb1);
+    gl.framebufferTexture2D(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex_2d, 1);
+    if (gl.checkFramebufferStatus(gl.DRAW_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE) {
+        testFailed("Draw framebuffer incomplete. Fail reason: " + wtu.glEnumToString(gl, status));
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: status is undefined.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    gl.texImage2D(gl.TEXTURE_CUBE_MAP_NEGATIVE_Y, 0, gl.RGBA8, width, height, 0, gl.RGBA, gl.UNSIGNED_BYTE, m_data0);
+    gl.texImage2D(gl.TEXTURE_CUBE_MAP_POSITIVE_Z, 0, gl.RGBA8, width, height, 0, gl.RGBA, gl.UNSIGNED_BYTE, m_data0);
+    gl.texImage2D(gl.TEXTURE_CUBE_MAP_NEGATIVE_Z, 0, gl.RGBA8, width, height, 0, gl.RGBA, gl.UNSIGNED_BYTE, m_data0);
+    gl.bindFramebuffer(gl.READ_FRAMEBUFFER, fb0);
+    gl.framebufferTexture2D(gl.READ_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_CUBE_MAP_POSITIVE_X, tex_cube_map, 0);
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fb1);
+    gl.framebufferTexture2D(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_CUBE_MAP_NEGATIVE_X, tex_cube_map, 0);
+    if ((gl.checkFramebufferStatus(gl.READ_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE) ||
+        (gl.checkFramebufferStatus(gl.DRAW_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE)) {
+        testFailed("Framebuffer incomplete. Fail reason: " + wtu.glEnumToString(gl, status));
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: status is undefined. Also, here you have two status.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    gl.texImage3D(gl.TEXTURE_2D_ARRAY, 0, gl.RGBA8, width, height, depth, 0, gl.RGBA, gl.UNSIGNED_BYTE, m_data1);
+    gl.bindFramebuffer(gl.READ_FRAMEBUFFER, fb0);
+    var level = 0, layer = 0;
+    gl.framebufferTextureLayer(gl.READ_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, tex_2d_array, level, layer);
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fb1);
+    layer = 1;
+    gl.framebufferTextureLayer(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, tex_2d_array, level, layer);
+    if ((gl.checkFramebufferStatus(gl.READ_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE) ||
+        (gl.checkFramebufferStatus(gl.DRAW_FRAMEBUFFER) != gl.FRAMEBUFFER_COMPLETE)) {
+        testFailed("Framebuffer incomplete. Fail reason: " + wtu.glEnumToString(gl, status));
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: status is undefined. You have two status.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+var gl = wtu.create3DContext("example", undefined, 2);
+
+function blit_framebuffer_test() {
+    var width = 8;
+    var height = 8;
+
+    // Create read fbo and its color attachment.
+    var tex_2d = gl.createTexture();
+    gl.bindTexture(gl.TEXTURE_2D, tex_2d);
+    var m_data0 = new Uint8Array(width * height * 4);
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: Because you are not verifying blit results, you really don't need the data.  Images will just be initialized to zero. This will simplify the test a bit.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        return;
+    }
+    gl.blitFramebuffer(0, 0, 2, 2, 0, 0, 2, 2, gl.COLOR_BUFFER_BIT | gl.COLOR_BUFFER_BIT, gl.NEAREST);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "blitFramebuffer should succeed if read/draw buffer has the same CUBE_MAP image with different faces.");
+
+    // The same 3D/2D_ARRAY image in read/draw buffer, but different layers are bound to read/draw buffer respectively.
+    var tex_2d_array = gl.createTexture();
+    gl.bindTexture(gl.TEXTURE_2D_ARRAY, tex_2d_array);
+    var depth = 2;
+    var m_data1 = new Uint8Array(width * height * depth * 4);
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: Same here, m_data1 is really unnecessary in this test.

**Comment by zhenyao** [comment]
Thanks for the thorough tests.  Some minor issues, otherwise it looks great.

File: specs/latest/2.0/index.html (hunk 1 above)
Comment (Richard-Yunchao) [inline specs/latest/2.0/index.html]: OK. I just saw that the gles3 man page is not correct. Developer man refer to that man page. 
I have removed this.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html (hunk 2 above)
Comment (Richard-Yunchao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: You are correct. The checkFramebufferStatus should pass, status info is useless for most cases. So remove it.

File: sdk/tests/conformance2/rendering/blitframebuffer-test.html (hunk 8 above)
Comment (Richard-Yunchao) [inline sdk/tests/conformance2/rendering/blitframebuffer-test.html]: That's true.

**Comment by Richard-Yunchao** [comment]
Thanks for your review, @zhenyao . Code has been updated accordingly. Please take another look. Thanks a lot!

**Comment by zhenyao** [comment]
lgtm
